### PR TITLE
adding zetasql-helper code

### PR DESCRIPTION
**Pull request by franklinWhaite**
This project is meant to be use as a starting point for developers interested in analyzing BigQuery SQL with [ZetaSQL](https://github.com/google/zetasql).

User should be able to base themselves of this implementation as it deals with most of the main tasks related to getting
started with ZetaSQL

### CODE REVIEW

**Comment by prathapreddy123** [comment]
@ppaglilla for review

**Review by ppaglilla — COMMENTED** [review]
Reviewed the PR! LGTM!

The only thing to point out is that we have the `target/` directory commited, which contains compiled code. We should remove that.

**Comment by franklinWhaite** [comment]
removed target thanks

File: tools/zetasql-helper/src/main/java/com/pso/bigquery/optimization/catalog/BigQueryTableSpec.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+ * limitations under the License.
+ */
+package com.pso.bigquery.optimization.catalog;
+
+import com.google.cloud.bigquery.Table;
+import com.google.cloud.bigquery.TableId;
+
+
+// Dataclass representing a BigQuery tables
+public class BigQueryTableSpec {
```
[the remark this answers is not in the document]
Comment (franklinWhaite) [reply tools/zetasql-helper/src/main/java/com/pso/bigquery/optimization/catalog/BigQueryTableSpec.java:23]: Makes sense, we can refactor

File: tools/zetasql-helper/src/main/java/com/pso/bigquery/optimization/analysis/ParsingUtils.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.pso.bigquery.optimization.analysis;
+
+import com.google.zetasql.ParseResumeLocation;
+
+public class ParsingUtils {
```
[the remark this answers is not in the document]
Comment (franklinWhaite) [reply tools/zetasql-helper/src/main/java/com/pso/bigquery/optimization/analysis/ParsingUtils.java:20]: Makes sense, will do. We will move the hasNextStatement method to the QueryAnalyzer

File: tools/zetasql-helper/src/main/java/com/pso/bigquery/optimization/analysis/QueryAnalysisResult.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    public List<BigQueryTableSpec> getReferencedTables() {
+        return referencedTables;
+    }
+
+    public JsonElement getQueryStructure() {
+        return queryStructure;
+    }
+
+    public String getQueryStructureAsString() {
+        return new Gson().toJson(this.queryStructure);
```
[the remark this answers is not in the document]
Comment (franklinWhaite) [reply tools/zetasql-helper/src/main/java/com/pso/bigquery/optimization/analysis/QueryAnalysisResult.java:59]: Makes sense, will do.

File: tools/zetasql-helper/src/main/java/com/pso/bigquery/optimization/catalog/BigQueryTableSpec.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (franklinWhaite) [reply tools/zetasql-helper/src/main/java/com/pso/bigquery/optimization/catalog/BigQueryTableSpec.java]: done

File: tools/zetasql-helper/src/main/java/com/pso/bigquery/optimization/catalog/CatalogDuplicateDatasetNameException.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.pso.bigquery.optimization.catalog;
+
+public class CatalogDuplicateDatasetNameException extends RuntimeException{
```
[the remark this answers is not in the document]
Comment (franklinWhaite) [reply tools/zetasql-helper/src/main/java/com/pso/bigquery/optimization/catalog/CatalogDuplicateDatasetNameException.java]: done

File: tools/zetasql-helper/src/main/java/com/pso/bigquery/optimization/BuildCatalogForProjectAndAnalyzeJoins.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+            + "FROM \n"
+            + "  `MY_PROJECT.MY_DATASET.test_table_1` t1\n"
+            + "LEFT JOIN\n"
+            + "  `MY_PROJECT.MY_DATASET.test_table_2` t2 ON t1.unique_key=t2.unique_key\n"
+            + "WHERE\n"
+            + " t1.col2 is not null\n"
+            + " AND t2.col2 is not null\n";
+
+        // Create Catalog and add all datasets in a project
+        SimpleCatalog catalog = CatalogUtils.createCatalogForProject(PROJECT_ID);
```
[the remark this answers is not in the document]
Comment (franklinWhaite) [reply tools/zetasql-helper/src/main/java/com/pso/bigquery/optimization/BuildCatalogForProjectAndAnalyzeJoins.java:47]: done

File: tools/zetasql-helper/src/main/java/com/pso/bigquery/optimization/analysis/ParsingUtils.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (franklinWhaite) [reply tools/zetasql-helper/src/main/java/com/pso/bigquery/optimization/analysis/ParsingUtils.java]: done

File: tools/zetasql-helper/src/main/java/com/pso/bigquery/optimization/analysis/QueryAnalysisResult.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (franklinWhaite) [reply tools/zetasql-helper/src/main/java/com/pso/bigquery/optimization/analysis/QueryAnalysisResult.java]: done

File: tools/zetasql-helper/src/main/java/com/pso/bigquery/optimization/analysis/QueryAnalyzer.java [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+import com.pso.bigquery.optimization.catalog.CatalogUtils;
+import com.pso.bigquery.optimization.catalog.*;
+import io.vavr.collection.Seq;
+import io.vavr.control.Try;
+import org.apache.commons.lang3.StringUtils;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+// The QueryAnalyzer parses a BigQuery job using ZetaSQL and
```
[the remark this answers is not in the document]
Comment (franklinWhaite) [reply tools/zetasql-helper/src/main/java/com/pso/bigquery/optimization/analysis/QueryAnalyzer.java]: done

File: tools/zetasql-helper/src/main/java/com/pso/bigquery/optimization/analysis/QueryAnalyzer.java [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+// extracts information from it.
+// It returns a QueryAnalysisResult with findings if the job
+// is successfully parsed with ZetaSQL.
+public class QueryAnalyzer {
+
+    private final BigQueryTableService bqTableService;
+
+    public QueryAnalyzer() {
+        this.bqTableService = null;
```
[the remark this answers is not in the document]
Comment (franklinWhaite) [reply tools/zetasql-helper/src/main/java/com/pso/bigquery/optimization/analysis/QueryAnalyzer.java]: done

File: tools/zetasql-helper/src/main/java/com/pso/bigquery/optimization/analysis/QueryAnalyzer.java [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+        );
+
+        if (tryUpdateCatalog.isFailure()) {
+            return Try.failure(tryUpdateCatalog.getCause());
+        }
+
+        while (ParsingUtils.hasNextStatement(parseResumeLocation)) {
+            Try<ResolvedStatement> tryParsedStatement = Try.of(() ->
+                    Analyzer.analyzeNextStatement(
+                            parseResumeLocation, this.getAnalyzerOptions(), catalog
```
[the remark this answers is not in the document]
Comment (franklinWhaite) [reply tools/zetasql-helper/src/main/java/com/pso/bigquery/optimization/analysis/QueryAnalyzer.java]: done

File: tools/zetasql-helper/src/main/java/com/pso/bigquery/optimization/analysis/QueryAnalyzer.java [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
```
[the remark this answers is not in the document]
Comment (franklinWhaite) [reply tools/zetasql-helper/src/main/java/com/pso/bigquery/optimization/analysis/QueryAnalyzer.java]: done

File: tools/zetasql-helper/src/main/java/com/pso/bigquery/optimization/analysis/visitors/BaseAnalyzerVisitor.java [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+                .getColumnDefinitionList()
+                .stream()
+                .map(definition -> new SimpleColumn(tableName, definition.getName(), definition.getType()))
+                .collect(Collectors.toList());
+    }
+
+    @Override
+    public void visit(ResolvedCreateTableStmt createTableStmt) {
+        List<String> tableNamePath = createTableStmt.getNamePath();
+        String tableName = tableNamePath.get(tableNamePath.size() - 1);
```
[the remark this answers is not in the document]
Comment (franklinWhaite) [reply tools/zetasql-helper/src/main/java/com/pso/bigquery/optimization/analysis/visitors/BaseAnalyzerVisitor.java:74]: done

**Comment by franklinWhaite** [comment]
Applied formatting, hope that's covered now!

**Review by prathapreddy123 — APPROVED** [review]

**Pull request merged**